contracts/rollover/CrossCurrencyRollover.sol

```diff
@@ -0,0 +1,609 @@
+     * @param recipient                 Address receiving the output token
+     *
+     * @return amountOut                The amount of tokenOut received.
+     */
+    function _swapExactInputSingle(
+        address tokenIn,
+        address tokenOut,
+        uint256 amountIn,
+        uint256 amountOutMinimum,
+        uint24 fee,
+        address recipient
+    ) internal returns (uint256 amountOut) {
+        // approve the UniswapV3 router to spend tokenIn
+        IERC20(tokenIn).safeApprove(address(swapRouter), amountIn);
+
+        // setting sqrtPriceLimitX96 to zero makes the parameter inactive.
+        // This parameter sets a boundary on the pool's swap price. It defines the
+        // worst acceptable price before the transaction reverts to allow for partial
+        // swaps, hence the value here is zero as we do not want to accept partial
+        // swaps.
+        ISwapRouter.ExactInputSingleParams memory params =
+            ISwapRouter.ExactInputSingleParams({
+                tokenIn: tokenIn,
+                tokenOut: tokenOut,
+                fee: fee,
+                recipient: recipient,
+                deadline: block.timestamp,
+                amountIn: amountIn,
+                amountOutMinimum: amountOutMinimum,
+                sqrtPriceLimitX96: 0
+            });
+
+        // execute the swap
+        amountOut = swapRouter.exactInputSingle(params);
+    }
+
+    // ======================================= LOAN REPAYMENT ===========================================
+    /**
+     * @notice This function sets up the old loan repayment data and triggers the repayment swap process
+     *         for the necessary funds to be collected from the new lender and swapped to the original
+     *         loan currency before being applied towards the repayment of the old loan.
+     *
+     * @param oldLoanId                 The ID of the original loan to be migrated.
+     * @param borrower_                 The address of the borrower.
+     * @param lender                    The address of the new lender.
+     * @param oldLoanCurrency           The currency of the old loan.
+     * @param newLoanTerms              The terms of the new currency loan.
+     * @param _amounts                  The rollover amounts.
+     * @param repayAmount               The principal and interest owed on the old loan.
+     * @param swapParams                Parameters for the currency swap, including
+     *                                  minimum swap amount out and pool fee tier.
+     */
+    function _initiateRepayment(
+        uint256 oldLoanId,
+        address borrower_,
+        address lender,
+        address oldLoanCurrency,
+        LoanLibrary.LoanTerms memory newLoanTerms,
+        OriginationLibrary.RolloverAmounts memory _amounts,
+        uint256 repayAmount,
+        OriginationLibrary.SwapParameters memory swapParams
+    ) internal {
+        // cache borrower address
+        borrower = borrower_;
```

Comment:
What's the effect achieved by caching the borrower address and then resetting later? Seems like a remnant from the security check we did for the flash loan flow